server/bleep/src/semantic/embedder.rs

```diff
@@ -65,89 +59,258 @@ pub struct EmbedChunk {
+            // the higher this number, the more vram we use, currently we use ~2G per session. this
+            // can be fixed by disabling scratch buffers in ggml, bert has no use for this.
+            let session_count = 3;
+
+            info!(%session_count, "spawned inference sessions");
+
+            let sessions = (0..session_count)
+                .map(|_| {
+                    model.start_session(llm::InferenceSessionConfig {
+                        ..Default::default()
+                    })
+                })
+                .map(tokio::sync::Mutex::new)
+                .map(Arc::new)
+                .collect();
+
+            // this tokenizer is used for chunking - do not pad or truncate chunks
+            let mut tokenizer =
+                Tokenizer::from_file(model_dir.join("ggml").join("tokenizer.json")).unwrap();
+            let _ = tokenizer.with_padding(None).with_truncation(None);
+
+            Ok(Self {
+                model,
+                sessions,
+                tokenizer,
+                permits: Arc::new(tokio::sync::Semaphore::new(session_count)),
+            })
+        }
     }
 
-    async fn batch_embed(&self, log: Vec<&str>) -> anyhow::Result<Vec<Embedding>> {
-        log.into_iter()
-            .map(|entry| tokio::task::block_in_place(|| self.embed(entry)))
-            .collect::<anyhow::Result<Vec<Embedding>>>()
+    #[async_trait]
+    impl Embedder for LocalEmbedder {
+        async fn embed(&self, sequence: &str) -> anyhow::Result<Embedding> {
+            let mut output_request = llm::OutputRequest {
+                all_logits: None,
+                embeddings: Some(Vec::new()),
+            };
+            let vocab = self.model.tokenizer();
+            let beginning_of_sentence = true;
+            let query_token_ids = vocab
+                .tokenize(sequence, beginning_of_sentence)
+                .unwrap()
+                .iter()
+                .map(|(_, tok)| *tok)
+                .collect::<Vec<_>>();
+
+            if let Ok(_permit) = self.permits.acquire().await {
+                for s in &self.sessions {
+                    if let Ok(mut session) = s.try_lock() {
+                        self.model
+                            .evaluate(&mut session, &query_token_ids, &mut output_request);
+                        return Ok(output_request.embeddings.unwrap());
+                    }
+                }
+            }
+
+            unreachable!();
+        }
+
+        fn tokenizer(&self) -> &Tokenizer {
+            &self.tokenizer
+        }
+
+        async fn batch_embed(&self, log: Vec<&str>) -> anyhow::Result<Vec<Embedding>> {
+            let mut output_request = llm::OutputRequest {
+                all_logits: None,
+                embeddings: Some(Vec::new()),
+            };
+            let vocab = self.model.tokenizer();
+            let beginning_of_sentence = true;
+            let query_token_ids = log
```

Comment:
This seems repeated from `embed`. Is it possible to express `batch_embed` in terms of `embed`, or parameterize the common bits to make it easy to spot the differences?